app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,1055 @@
+        File favoritesDir =
+                new File(LocalDatabasePaths.favoritesFolder());
+        FileIO.deleteDir(favoritesDir);
+        clearMemory();
+    }
+
+    /**
+     * This method finds a card by id in the cards array and returns the
+     * first index that reference a card with the same card id.
+     *
+     * @param cardId the id of the card
+     * @return the index of the card in cards, or -1 if it doesn't find it.
+     */
+    private int findCardById(String cardId) {
+        for (int i = 0; i < this.cards.size(); ++i) {
+            Card card = this.cards.get(i);
+            if (Objects.requireNonNull(card.getId()).equals(cardId)) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+
+    /**
+     * Removes a card with id cardId. The user associated with this card is
+     * removed only if it isn't referenced by any other card.
+     *
+     * @param cardId the id of the card
+     */
+    public void removeCard(String cardId) {
+        String pathToCard =
+                LocalDatabasePaths.cardFolder(getCurrentUser().getUserId(),
+                        cardId);
+        FileIO.deleteDir(new File(pathToCard));
+
+        int cardIdx = findCardById(cardId);
+
+        if (cardIdx == -1) return;
+
+        Card card = this.cards.get(cardIdx);
+
+        String adId = card.getAdId();
+        String userId = card.getUserId();
+
+        this.idsToAd.remove(adId);
+        this.cards.remove(cardIdx);
+        this.adIdsToPanoramas.remove(adId);
+
+        boolean isUserUsed = false;
+        for (int i = 0; i < this.cards.size() && !isUserUsed; ++i) {
+            Card curCard = this.cards.get(i);
+
+            if (curCard.getUserId().equals(userId)) {
+                isUserUsed = true;
+            }
+        }
+
+        if (!isUserUsed) {
+            this.userIds.remove(userId);
+            this.idsToUser.remove(userId);
+            String userPath =
+                    LocalDatabasePaths.userFolder(getCurrentUser().getUserId(), userId);
+            FileIO.deleteDir(new File(userPath));
+        }
+    }
+
+    /**
+     * This sets the current user for the local database. The favorites are
+     * stored per user. It is useful in the case where two accounts are used
+     * on the same phone.
+     * <p>
+     * Also, the last currentUser is stored on disk so that when the user
+     * goes into the app while offline, the app can "guess" who this is and
+     * thus retrieve the correct favorite data.
+     * <p>
+     * This method should only be called when the app is online.
+     *
+     * @param currentUser    the current user
+     * @param profilePicLoad the function that will perform the writing of
+     *                       the profile picture of the user on disk
+     * @return a completable future that indicates if the operation succeeded
+     * or not.
+     */
+    public CompletableFuture<Void> setCurrentUser(User currentUser,
```

Comment:
I replied in more details in my reply to your first comment (summary of the review). But just to summarize it here, I will have a look at the APIs and see if it is interesting to have asynchronous disk operations for the kind of file sizes we are dealing with. If it happens to not be interesting performance wise or too complicated, I will just drop the CompletableFuture.

Reply:
I sort of agree with @rovati, the futures bring complexity to the api also in the client code since he handle it. We should benchmark this in order to choose but if the app freezes for 500ms when setting the user I don't know if it is worth it.